resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

```diff
@@ -0,0 +1,218 @@
+          val otherKubernetesResources = currentDriverSpec.otherKubernetesResources
+          addDriverOwnerReference(createdDriverPod, otherKubernetesResources)
+          kubernetesClient.resourceList(otherKubernetesResources: _*).createOrReplace()
+        }
+      } catch {
+        case NonFatal(e) =>
+          kubernetesClient.pods().delete(createdDriverPod)
+          throw e
+      }
+
+      if (waitForAppCompletion) {
+        logInfo(s"Waiting for application $appName to finish...")
+        loggingPodStatusWatcher.awaitCompletion()
+        logInfo(s"Application $appName finished.")
+      } else {
+        logInfo(s"Deployed Spark application $appName into Kubernetes.")
+      }
+    }
+  }
+
+  // Add a OwnerReference to the given resources making the driver pod an owner of them so when
+  // the driver pod is deleted, the resources are garbage collected.
+  private def addDriverOwnerReference(driverPod: Pod, resources: Seq[HasMetadata]): Unit = {
+    val driverPodOwnerReference = new OwnerReferenceBuilder()
+      .withName(driverPod.getMetadata.getName)
+      .withApiVersion(driverPod.getApiVersion)
+      .withUid(driverPod.getMetadata.getUid)
+      .withKind(driverPod.getKind)
+      .withController(true)
+      .build()
+    resources.foreach { resource =>
+      val originalMetadata = resource.getMetadata
+      originalMetadata.setOwnerReferences(Collections.singletonList(driverPodOwnerReference))
+    }
+  }
+}
+
+private[spark] object Client {
+
+  def run(sparkConf: SparkConf, clientArguments: ClientArguments): Unit = {
+    val namespace = sparkConf.get(KUBERNETES_NAMESPACE)
+    val kubernetesAppId = s"spark-${UUID.randomUUID().toString.replaceAll("-", "")}"
+    val launchTime = System.currentTimeMillis()
+    val waitForAppCompletion = sparkConf.get(WAIT_FOR_APP_COMPLETION)
+    val appName = sparkConf.getOption("spark.app.name").getOrElse("spark")
+    val master = getK8sMasterUrl(sparkConf.get("spark.master"))
+    val loggingInterval = Option(sparkConf.get(REPORT_INTERVAL)).filter(_ => waitForAppCompletion)
+
+    val loggingPodStatusWatcher = new LoggingPodStatusWatcherImpl(
+      kubernetesAppId, loggingInterval)
+
+    val configurationStepsOrchestrator = new DriverConfigurationStepsOrchestrator(
+      namespace,
+      kubernetesAppId,
+      launchTime,
+      clientArguments.mainAppResource,
+      appName,
+      clientArguments.mainClass,
+      clientArguments.driverArgs,
+      sparkConf)
+
+    Utils.tryWithResource(KubernetesClientFactory.createKubernetesClient(
+      master,
+      Some(namespace),
+      KUBERNETES_AUTH_SUBMISSION_CONF_PREFIX,
+      sparkConf,
+      None,
+      None)) {
+      kubernetesClient =>
+        new Client(
+          configurationStepsOrchestrator.getAllConfigurationSteps(),
+          sparkConf,
+          kubernetesClient,
+          waitForAppCompletion,
+          appName,
+          loggingPodStatusWatcher).run()
```

Comment:
I almost missed `.run()` at the end. Maybe we can use a`val` here:
```
kubernetesClient =>
        val sparkClient = new Client(
           configurationStepsOrchestrator.getAllConfigurationSteps(),
           sparkConf,
           kubernetesClient,
           waitForAppCompletion,
           appName,
           loggingPodStatusWatcher)
        sparkClient.run()
```

Reply:
Done.